src/systems/filecoin_mining/storage_mining/storage_miner_actor.go

```diff
@@ -1,21 +1,355 @@
+// into Cleared State which means deleting the Sector from state
+// remove SectorNumber from all states on chain
+func (sm *StorageMinerActor_I) clearSector(sectorNo sector.SectorNumber) {
+	delete(sm.Sectors(), sectorNo)
+	delete(sm.SectorStates(), sectorNo)
+	sm.ProvingSet_.Remove(sectorNo)
+}
+
+// move Sector from Committed/Recovering into Active State
+// reset FaultCount to zero
+func (sm *StorageMinerActor_I) activateSector(sectorNo sector.SectorNumber) {
+	sm.SectorStates()[sectorNo] = SectorActive()
+}
+
+// move Sector from Active/Committed/Recovering into Failing State
+// and increment FaultCount
+func (sm *StorageMinerActor_I) failSector(sectorNo sector.SectorNumber) {
+	newFaultCount := sm.SectorStates()[sectorNo].FaultCount + 1
+	sm.ProvingSet_.Remove(sectorNo)
+	sm.SectorStates()[sectorNo] = SectorFailing(newFaultCount)
+}
+
+// increment FaultCount and check if greater than MaxFaultCount
+// move from Failing to Cleared State if so
+func (sm *StorageMinerActor_I) checkMFC(sectorNo sector.SectorNumber) {
+	newFaultCount := sm.SectorStates()[sectorNo].FaultCount + 1
+	if newFaultCount > MAX_CONSECUTIVE_FAULT_COUNT {
+		// TODO: heavy penalization: slash pledge collateral and delete sector
+		// TODO: SendMessage(SPA.SlashPledgeCollateral)
+		sm.clearSector(sectorNo)
 	} else {
-		newSector := append(currSector, onChainInfo)
-		sma.Sectors()[onChainInfo.SectorNumber()] = newSector
+		// increment FaultCount
+		// TODO: SendMessage(sma.SlashStorageDealCollateral)
+		sm.SectorStates()[sectorNo] = SectorFailing(newFaultCount)
+	}
+}
+
+// Decision is to currently account for power based on sector
+// with at least one active deals and deals cannot be updated
+// an alternative proposal is to account for power based on active deals
+// an improvement proposal is to allow storage deal update in a sector
+
+// TODO: decide whether declared faults sectors should be
+// penalized in the same way as undeclared sectors and how
+
+// SubmitPoSt Workflow:
+// - Verify PoSt Submission
+// - Process ProvingSet.GetOnes()
+//   - State Transitions
+//     - Committed -> Active and credit power
+//     - Recovering -> Active and credit power
+//   - Process Active Sectors (pay miners)
+// - Process ProvingSet.GetZeros()
+//	   - increment FaultCount
+//     - clear Sector and slash pledge collateral if count > MAX_CONSECUTIVE_FAULT_COUNT
+// - Process Expired Sectors (settle deals and return storage collateral to miners)
+//     - State Transition
+//       - Failing / Recovering / Active / Committed -> Cleared
+//     - Remove SectorNumber from Sectors, SectorStates, ProvingSet
+func (sm *StorageMinerActor_I) SubmitPoSt(postSubmission poster.PoStSubmission) {
+	// Verify correct PoSt Submission
+	isPoStVerified := sm.verifyPoStSubmission(postSubmission)
+	if !isPoStVerified {
+		// no state transition, just error out and miner should submitPoSt again
+		// TODO: proper failure
+		panic("TODO")
```

Comment:
yep, correct. should note here (as a comment):

```go
// miners get to try again. we do not need to address the failure case here
// it will be addressed via a `CronActor` calling `StoragePowerActor`
```
or something like this

Reply:
Should it not return an error code, pay gas fees etc? I don't think actor code should have a panic anywhere.